### PR TITLE
pkg/proc: Refactor process post initialization

### PR DESCRIPTION
**Pull request by derekparker**
This patch is a slight refactor to share more code used for generic
process initialization. There will always be OS/backend specific
initialization, but as much as can be shared should be to prevent
duplicating of any logic (setting internal breakpoints, loading bininfo,
etc).

### CODE REVIEW

File: pkg/proc/interface.go [hunk 1]
```diff
@@ -105,13 +108,17 @@ type ProcessManipulation interface {
...
+	WriteBreakpointFn() WriteBreakpointFn
 	ClearBreakpoint(addr uint64) (*Breakpoint, error)
 	ClearInternalBreakpoints() error
 }
 
 // CommonProcess contains fields used by this package, common to all
 // implementations of the Process interface.
 type CommonProcess struct {
+	// Path is the path on disk to the executable being debugged.
+	Path string
```
Comment (aarzilli) [inline pkg/proc/interface.go]: This seems redundant with the same field in BinaryInfo. Given that it only exists to pass this information to PostInitializationSetup I think it's fine if it's a parameter to that function.

File: pkg/proc/gdbserial/gdbserver.go [hunk 2]
```diff
@@ -545,10 +448,89 @@ func LLDBAttach(pid int, path string, debugInfoDirs []string) (*Process, error)
...
+	// overwrite some existing memory to store the MOV.
+	if addr, err := p.conn.allocMemory(256); err == nil {
+		if _, err := p.conn.writeMemory(uintptr(addr), p.loadGInstr()); err == nil {
+			p.loadGInstrAddr = addr
+		}
+	}
+
+	err = p.updateThreadList(&threadUpdater{p: p})
+	if err != nil {
+		p.bi.Close()
```
Comment (aarzilli) [inline pkg/proc/gdbserial/gdbserver.go]: Not sure why you have removed the calls to conn.Close here and elsewhere and strongly suspect it's wrong.

File: pkg/proc/interface.go [hunk 3]
```diff
@@ -105,13 +108,17 @@ type ProcessManipulation interface {
 type BreakpointManipulation interface {
 	Breakpoints() *BreakpointMap
 	SetBreakpoint(addr uint64, kind BreakpointKind, cond ast.Expr) (*Breakpoint, error)
+	WriteBreakpointFn() WriteBreakpointFn
```
[the remark this answers is not in the document]
Comment (derekparker) [reply pkg/proc/interface.go]: Changed.

File: pkg/proc/interface.go [hunk 4]
```diff
@@ -86,10 +87,12 @@ type ThreadInfo interface {
 // GoroutineInfo is an interface for getting information on running goroutines.
 type GoroutineInfo interface {
 	SelectedGoroutine() *G
+	SetSelectedGoroutine(*G)
 }
 
 // ProcessManipulation is an interface for changing the execution state of a process.
 type ProcessManipulation interface {
+	Initialize([]string) error
```
[the remark this answers is not in the document]
Comment (derekparker) [reply pkg/proc/interface.go]: Agreed, changed.

File: pkg/proc/interface.go (hunk 1 above)
Comment (derekparker) [reply pkg/proc/interface.go]: Agreed, changed.

File: pkg/proc/gdbserial/gdbserver.go (hunk 2 above)
Comment (derekparker) [reply pkg/proc/gdbserial/gdbserver.go]: Fixed.

File: pkg/proc/proc.go [hunk 5]
```diff
@@ -41,6 +41,38 @@ func (pe ProcessDetachedError) Error() string {
...
+	if err != nil {
+		return err
+	}
+
+	g, _ := GetG(p.CurrentThread())
+	p.SetSelectedGoroutine(g)
+
+	createUnrecoveredPanicBreakpoint(p, writeBreakpoint)
+
+	panicpc, err := FindFunctionLocation(p, "runtime.startpanic", true, 0)
```
[the remark this answers is not in the document]
Comment (derekparker) [reply pkg/proc/proc.go]: Fixed.

**Pull request merged**